How many late days are you using for this assignment? 0 

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import re

# Part 1: Load the dataset

In [2]:
# Load the given datasets
KDS_numerical_dataset_1_QL = pd.read_csv('chronic_kidney_disease_numerical.csv')
KDS_categorical_dataset_2_QL = pd.read_csv('chronic_kidney_disease_categorical.csv')

# Print the data
display("numerical data set",KDS_numerical_dataset_1_QL)
display("categorical data set",KDS_categorical_dataset_2_QL)

'numerical data set'

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


'categorical data set'

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

According to the data set, there are about 25 features. They are: 
1) age (numerical): patient's age information 
2) blood pressure (numerical) measures the pressure of blood
3) sod (numerical) calculates density of sodium in 1 litter of blood,
4) pc (numerical)  refers to pus cells, which are responsible for collecting dead cells
5) su (categorical) refers to sugar in blood
6) al (categorical) refers to albumin
7) sg (categotical) is specific gravity. Sg is a measure of urine concentration.
8) bu (numerical) describes blood urea, unit is mgs/dl. bu measures amount of urea nitrogen in blood.
9) sc (numerical) measures how well the kidney function is based on the density of creatinine in 1 dl of blood.
10) wbcc(numerical) counts total number of white blood cells in 1 cmm of blood
11) rbcc(numerical) counts total number of red blood cells in 1 cmm of blood
12) htn(nominal, categorical): binary values to evaluate whether the patient is high blood pressure (hypertension) or not.
13) dm (nominal, categorical): refers to diabetes mellitus, which indicates the presense of group of diseases that cause damage to the kidneys
14) pe(nominal, categorical): pedal edema indicates whether there are any symptoms such as swelling in the feet or ankles.
15) pcc(nominal, categorical): pus cell clumsps indicates whether white cells reponse to bateria.
16) rbc(nominal, categotical): indicates the presence of red blood cells in a sample of urine. 
17) bgr(numerica): measures amount of glucose (mgs) in 1 dl of blood at any time. 
18) ba(categorical): indicates the presense of bacteria in sample of urine or blood.
19) pot(numerical): measures the number of potassium (mEq) in 1 litter of blood. 
20) hemo(numerical): measures the number of hemoglobin in gms units.
21) pcv(numerical): calculates number of packed cell volume. The unit is unknown. 
22) ane(categorical): check if the kidneys are exposed with anemia disease. 
23) appet(categorical): check the patient has a desire to eat.
24) cad(categorical): check the presence of coronary artery disease.
 
The label used in this data set is 'Target'. 'Target' is a binary (categorical) indicating whether patient's kidneys are in serious condition or not. There are two values: ckd: high risk, and not ckd is normal Answer:

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_rows_bool = KDS_numerical_dataset_1_QL.duplicated(subset= 'unique_id').any()
if(duplicate_rows_bool):
    display(f"number of duplicated rows of numerical dataset is:\
     {KDS_numerical_dataset_1_QL.duplicated(subset = 'unique_id').sum()}")
# Drop these duplicate rows
    KDS_numerical_dataset_1_QL.drop_duplicates(subset= 'unique_id', inplace=True)
    KDS_numerical_dataset_1_QL.reset_index(drop = True, inplace = True)
display('numerical data set after dropping duplicate rows',KDS_numerical_dataset_1_QL)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_rows_bool_2 = KDS_categorical_dataset_2_QL.duplicated(subset= 'unique_id').any()
if(duplicate_rows_bool_2):
    display(f"number of duplicated rows of categorical dataset is:\
    {KDS_categorical_dataset_2_QL.duplicated(subset = 'unique_id').sum()}")
    display('display duplicate rows ',\
    KDS_categorical_dataset_2_QL[KDS_categorical_dataset_2_QL.duplicated(subset='unique_id')])
    KDS_categorical_dataset_2_QL.drop_duplicates(subset= 'unique_id',inplace=True)
    KDS_categorical_dataset_2_QL.reset_index(drop =True, inplace = True)
display('categorical data set after dropping duplicate rows',KDS_categorical_dataset_2_QL)



'number of duplicated rows of numerical dataset is:     30'

'numerical data set after dropping duplicate rows'

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


'number of duplicated rows of categorical dataset is:    30'

'display duplicate rows '

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
231,856244,0.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,poor,no,no,ckd
401,628271,2.0,1.0,abnormal,normal,notpresent,notpresent,yes,yes,no,poor,no,yes,ckd
402,158929,1.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,ckd
403,240975,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
404,441170,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
405,337834,4.0,0.0,normal,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
406,118191,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
407,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
408,302906,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,no,yes,ckd
409,938027,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


'categorical data set after dropping duplicate rows'

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
396,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd
398,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
Not all unique ids are present in both datasets. Since the size of categorical data set is greater than that of numerical data set - 10 rows and 4 columns -.
To examine the missing rows in the combined table, I slightly modified the merge function by setting the 'how' parameter to 'outer'. By default, the merge function performs an 'inner' join, which automatically excludes non-matching rows from the combined dataset. After reconfiguring, the missing rows are represented by NaN values in columns without corresponding data.

For example:

unique_id age	bp	bgr	bu	sc	sod	pot	hemo pcv	...	 pcc	    ba	       htn	dm	cad	appet   pe	ane

208034	  NaN	NaN	NaN	NaN	NaN	NaN	NaN	NaN	 NaN	...	notpresent	notpresent	yes	yes	yes	good	yes	 no	

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_data_set_by_merging_QL = pd.merge(KDS_numerical_dataset_1_QL,KDS_categorical_dataset_2_QL,\
                                 on = 'unique_id', how = 'outer')
#Print the combined dataset
display('name of columns',combined_data_set_by_merging_QL.columns)
display(combined_data_set_by_merging_QL)


'name of columns'

Index(['unique_id', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo',
       'pcv', 'wbcc', 'rbcc', 'al', 'su', 'rbc', 'pc', 'pcc', 'ba', 'htn',
       'dm', 'cad', 'appet', 'pe', 'ane', 'Target'],
      dtype='object')

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103570,24.0,80.0,125.0,NaN,NaN,136.0,3.5,15.4,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,104675,65.0,80.0,215.0,133.0,2.5,NaN,NaN,13.2,41.0,...,normal,present,notpresent,no,yes,no,good,no,no,ckd
3,105923,90.0,90.0,139.0,89.0,3.0,140.0,4.1,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
4,107297,60.0,70.0,NaN,NaN,NaN,NaN,NaN,16.4,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd
396,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,996412,70.0,60.0,120.0,40.0,0.5,140.0,4.6,16.0,43.0,...,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
398,997629,35.0,NaN,99.0,30.0,0.5,135.0,4.9,15.4,48.0,...,normal,NaN,NaN,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value
the_percentage_of_missing_value_rows = combined_data_set_by_merging_QL.isnull().any(axis=1).mean()*(100)
# Print %
print(the_percentage_of_missing_value_rows,"%")
# Drop these rows from the dataset
combined_data_set_by_merging_QL.dropna(how='any',inplace=True)
combined_data_set_by_merging_QL.reset_index(drop = True, inplace = True)

# Print the Dataset
display('name of columns',combined_data_set_by_merging_QL.columns)
display(combined_data_set_by_merging_QL)

61.75000000000001 %


'name of columns'

Index(['unique_id', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo',
       'pcv', 'wbcc', 'rbcc', 'al', 'su', 'rbc', 'pc', 'pcc', 'ba', 'htn',
       'dm', 'cad', 'appet', 'pe', 'ane', 'Target'],
      dtype='object')

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
4,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,...,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
149,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
combined_data_set_by_merging_QL.sort_values(by= 'Target').reset_index(drop = True,inplace=True)
# Print the dataset
display(combined_data_set_by_merging_QL)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
4,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,...,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
149,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encode_dum_columns_QL_merge = pd.get_dummies(combined_data_set_by_merging_QL,\
                            columns=categorical_columns,\
                            prefix=categorical_columns,dtype=int)
# Print the dataset
display('name of columns',encode_dum_columns_QL_merge.columns)
display(encode_dum_columns_QL_merge)

'name of columns'

Index(['unique_id', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo',
       'pcv', 'wbcc', 'rbcc', 'al', 'su', 'rbc_abnormal', 'rbc_normal',
       'pc_abnormal', 'pc_normal', 'pcc_notpresent', 'pcc_present',
       'ba_notpresent', 'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes',
       'cad_no', 'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes',
       'ane_no', 'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,1,0,0,1,0,1,0,1,1,0
1,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,1,0,1,0,1,0,1,0,0,1
2,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,...,1,0,1,0,1,0,1,0,0,1
3,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,...,1,0,1,0,1,0,1,0,0,1
4,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,...,1,0,1,0,1,0,1,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,...,1,0,1,0,1,0,1,0,1,0
149,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,...,1,0,1,0,1,0,1,0,0,1
150,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,...,1,0,1,0,1,0,1,0,0,1
151,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,0,1,1,0,0,1,0,1,1,0


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: if the column only has 2 opposite status: this or that, then in the downstream modeling this should be equivalent. It is totally fine with one single column such as 'appet' which has values 0 and 1 indicating whether the patient has desire to eat or not. However, the single column solution should not be applied to cases where categorical variables could take more than 2 values since this will cause confusion and confliction. That is, with one single column, we can not determine which value is on the data set because we only have 2 indicators 0 and 1 using for all values.  

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [8]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
list_of_means = encode_dum_columns_QL_merge[numerical_columns].mean()
list_of_stds = encode_dum_columns_QL_merge[numerical_columns].std()
list_of_upper_bounds = list_of_means + 3*list_of_stds
list_of_lower_bounds = list_of_means - 3*list_of_stds
display(list_of_lower_bounds)
for value in numerical_columns:
    encode_dum_columns_QL_merge = encode_dum_columns_QL_merge[\
    (encode_dum_columns_QL_merge[value] >= list_of_lower_bounds[value])\
     & (encode_dum_columns_QL_merge[value] <= list_of_upper_bounds[value])].dropna()
# Print the dataset
encode_dum_columns_QL_merge.reset_index(drop = True, inplace = True)
display(encode_dum_columns_QL_merge)

age        2.598464
bp        40.966650
bgr      -64.439238
bu       -90.197571
sc        -6.837725
sod      116.126905
pot       -5.945258
hemo       5.082753
pcv       14.929232
wbcc   -1053.207280
rbcc       1.797915
dtype: float64

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,1,0,1,0,1,0,1,0,0,1
1,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,...,1,0,1,0,1,0,1,0,0,1
2,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,...,1,0,1,0,1,0,1,0,0,1
3,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,...,1,0,1,0,1,0,1,0,0,1
4,124304,69.0,70.0,264.0,87.0,2.7,130.0,4.0,12.5,37.0,...,0,1,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,...,1,0,1,0,1,0,1,0,1,0
130,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,...,1,0,1,0,1,0,1,0,0,1
131,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,...,1,0,1,0,1,0,1,0,0,1
132,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,0,1,1,0,0,1,0,1,1,0


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [9]:
# Normalize the all Numerical Attributes in the dataset.
for value in numerical_columns:
    encode_dum_columns_QL_merge[value] = (encode_dum_columns_QL_merge[value] - encode_dum_columns_QL_merge[value].min())/\
                                        (encode_dum_columns_QL_merge[value].max() - encode_dum_columns_QL_merge[value].min())
# Print the dataset
display(encode_dum_columns_QL_merge)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,109053,0.783784,0.00,0.154506,0.111111,0.037037,1.000000,0.114286,0.721311,0.684211,...,1,0,1,0,1,0,1,0,0,1
1,114717,0.472973,0.50,0.223176,0.098039,0.049383,0.600000,0.600000,0.942623,0.657895,...,1,0,1,0,1,0,1,0,0,1
2,118191,0.486486,0.25,0.098712,0.143791,0.061728,0.766667,0.514286,0.901639,0.710526,...,1,0,1,0,1,0,1,0,0,1
3,118596,0.918919,0.00,0.077253,0.261438,0.024691,0.900000,0.228571,0.950820,0.973684,...,1,0,1,0,1,0,1,0,0,1
4,124304,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,0.565574,0.552632,...,0,1,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,980291,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,0.605263,...,1,0,1,0,1,0,1,0,1,0
130,991031,0.324324,0.50,0.081545,0.209150,0.012346,0.633333,0.600000,0.909836,0.947368,...,1,0,1,0,1,0,1,0,0,1
131,994377,0.878378,0.25,0.047210,0.052288,0.049383,0.833333,0.600000,0.745902,0.736842,...,1,0,1,0,1,0,1,0,0,1
132,995177,0.851351,0.25,0.618026,0.562092,0.728395,0.000000,0.285714,0.311475,0.315789,...,0,1,1,0,0,1,0,1,1,0


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Yes, the unique_id column is not appropriate for modeling and predictions since unique_ids are just arbitrary numbers and do not contribute to the data set.

In [10]:
#Remove that column
encode_dum_columns_QL_merge = encode_dum_columns_QL_merge.drop(['unique_id'], axis=1)
# Print the dataset
display(encode_dum_columns_QL_merge)
top_15_rows = encode_dum_columns_QL_merge.iloc[0:15]
display(encode_dum_columns_QL_merge.columns)
print(top_15_rows.to_string(index=False))

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.783784,0.00,0.154506,0.111111,0.037037,1.000000,0.114286,0.721311,0.684211,0.306452,...,1,0,1,0,1,0,1,0,0,1
1,0.472973,0.50,0.223176,0.098039,0.049383,0.600000,0.600000,0.942623,0.657895,0.387097,...,1,0,1,0,1,0,1,0,0,1
2,0.486486,0.25,0.098712,0.143791,0.061728,0.766667,0.514286,0.901639,0.710526,0.225806,...,1,0,1,0,1,0,1,0,0,1
3,0.918919,0.00,0.077253,0.261438,0.024691,0.900000,0.228571,0.950820,0.973684,0.137097,...,1,0,1,0,1,0,1,0,0,1
4,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,0.565574,0.552632,0.427419,...,0,1,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,0.605263,0.290323,...,1,0,1,0,1,0,1,0,1,0
130,0.324324,0.50,0.081545,0.209150,0.012346,0.633333,0.600000,0.909836,0.947368,0.475806,...,1,0,1,0,1,0,1,0,0,1
131,0.878378,0.25,0.047210,0.052288,0.049383,0.833333,0.600000,0.745902,0.736842,0.443548,...,1,0,1,0,1,0,1,0,0,1
132,0.851351,0.25,0.618026,0.562092,0.728395,0.000000,0.285714,0.311475,0.315789,0.580645,...,0,1,1,0,0,1,0,1,1,0


Index(['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc',
       'rbcc', 'al', 'su', 'rbc_abnormal', 'rbc_normal', 'pc_abnormal',
       'pc_normal', 'pcc_notpresent', 'pcc_present', 'ba_notpresent',
       'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes', 'cad_no',
       'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes', 'ane_no',
       'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')

     age   bp      bgr       bu       sc      sod      pot     hemo      pcv     wbcc     rbcc  al  su  rbc_abnormal  rbc_normal  pc_abnormal  pc_normal  pcc_notpresent  pcc_present  ba_notpresent  ba_present  htn_no  htn_yes  dm_no  dm_yes  cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  Target_ckd  Target_notckd
0.783784 0.00 0.154506 0.111111 0.037037 1.000000 0.114286 0.721311 0.684211 0.306452 0.538462 0.0 0.0             0           1            0          1               1            0              1           0       1        0      1       0       1        0           1           0      1       0       1        0           0              1
0.472973 0.50 0.223176 0.098039 0.049383 0.600000 0.600000 0.942623 0.657895 0.387097 0.666667 0.0 0.0             0           1            0          1               1            0              1           0       1        0      1       0       1        0           1           0      1       0       1        0     

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file
encode_dum_columns_QL_merge.to_csv('csv_version_of_QL_data_set', index= False)
# Export the dataframe to a new json file
encode_dum_columns_QL_merge.to_json('json_version_of_QL_data_set', index= False)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

### Paste the markdown table here
### Paste the markdown table here
|     age |    bp |       bgr |        bu |        sc |       sod |       pot |      hemo |      pcv |      wbcc |      rbcc |   al |   su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|---------|-------|-----------|-----------|-----------|-----------|-----------|------------|-----------|-----------|-----------|------|------|--------------|------------|-------------|-----------|-----------------|-------------|----------------|------------|--------|---------|-------|--------|--------|---------|------------|-------------|-------|--------|--------|---------|-------------|-----------------|
| 0.783784| 0.00  | 0.154506  | 0.111111  | 0.037037  | 1.000000  | 0.114286  | 0.721311   | 0.684211  | 0.306452  | 0.538462  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.472973| 0.50  | 0.223176  | 0.098039  | 0.049383  | 0.600000  | 0.600000  | 0.942623   | 0.657895  | 0.387097  | 0.666667  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.486486| 0.25  | 0.098712  | 0.143791  | 0.061728  | 0.766667  | 0.514286  | 0.901639   | 0.710526  | 0.225806  | 0.692308  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.918919| 0.00  | 0.077253  | 0.261438  | 0.024691  | 0.900000  | 0.228571  | 0.950820   | 0.973684  | 0.137097  | 0.487179  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.851351| 0.25  | 0.832618  | 0.503268  | 0.283951  | 0.333333  | 0.314286  | 0.565574   | 0.552632  | 0.427419  | 0.384615  | 3.0  | 4.0  |            0 |          1 |           1 |         0 |               1 |           0 |              1 |          0 |      0 |       1 |     0 |      1 |      0 |       1 |          1 |          0 |     0 |      1 |      1 |       0 |          1  |               0 |
| 0.418919| 0.00  | 0.167382  | 0.241830  | 0.086420  | 0.700000  | 0.571429  | 0.770492   | 0.842105  | 0.217742  | 0.666667  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.567568| 0.50  | 0.270386  | 0.843137  | 1.000000  | 0.400000  | 0.742857  | 0.385246   | 0.526316  | 0.153226  | 0.358974  | 4.0  | 0.0  |            1 |          0 |           1 |         0 |               1 |           0 |              0 |          1 |      1 |       0 |     0 |      1 |      1 |       0 |          1 |          0 |     0 |      1 |      1 |       0 |          1  |               0 |
| 0.621622| 0.50  | 0.236052  | 0.078431  | 0.098765  | 0.633333  | 0.485714  | 0.893443   | 0.710526  | 0.032258  | 0.512821  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.310811| 0.50  | 0.055794  | 0.254902  | 0.061728  | 0.633333  | 0.114286  | 0.975410   | 0.631579  | 0.451613  | 0.538462  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.905405| 1.00  | 0.965665  | 0.522876  | 0.641975  | 0.666667  | 0.000000  | 0.295082   | 0.368421  | 0.217742  | 0.153846  | 3.0  | 2.0  |            1 |          0 |           1 |         0 |               0 |           1 |              1 |          0 |      0 |       1 |     0 |      1 |      0 |       1 |          0 |          1 |     1 |      0 |      1 |       0 |          1  |               0 |
| 0.837838| 0.00  | 0.236052  | 0.202614  | 0.086420  | 0.633333  | 0.257143  | 0.967213   | 0.894737  | 0.193548  | 0.897436  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.716216| 0.00  | 0.184549  | 0.084967  | 0.086420  | 0.633333  | 0.171429  | 0.795082   | 1.000000  | 0.177419  | 0.589744  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.567568| 0.00  | 0.180258  | 0.222222  | 0.098765  | 0.733333  | 0.571429  | 0.729508   | 0.736842  | 0.411290  | 0.974359  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.810811| 0.25  | 0.158798  | 0.039216  | 0.086420  | 0.666667  | 0.200000  | 0.655738   | 0.684211  | 0.540323  | 0.589744  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |
| 0.608108| 0.00  | 0.124464  | 0.183007  | 0.049383  | 0.500000  | 0.228571  | 0.606557   | 0.868421  | 0.322581  | 0.666667  | 0.0  | 0.0  |            0 |          1 |           0 |         1 |               1 |           0 |              1 |          0 |      1 |       0 |     1 |      0 |      1 |       0 |          1 |          0 |     1 |      0 |      1 |       0 |          0  |               1 |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

At first cs383GPT does not work well with large data set (larger than 10 queries)
- Secondly, I detect that cs383GPT makes a mistake at some values if I paste values directly from cvs or json files.
Thus, to mitigate data conversion hallucinations, I have to convert the dataframe into strings via built in function 'to_string'. By this way, the LLM can correctly match values to corresponding features.  

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [12]:
# Code to rename all the columns in the dataset

def change_to_full_name(column):
    for abbreviation in full_name_map:
        if re.match(rf'^{abbreviation}(\
        _normal|_abnormal|_yes|_no|_present|_notpresent|_good|_poor)?$', column):
            root_nae = full_name_map[abbreviation]
            if column.endswith('_yes'):
                return f"{root_nae} Yes"
            elif column.endswith('_no'):
                return f"{root_nae} No"
            elif column.endswith('_normal'):
                return f"{root_nae} Normal"
            elif column.endswith('_abnormal'):
                return f"{root_nae} Abnormal"
            elif column.endswith('_present'):
                return f"{root_nae} present"
            elif column.endswith('_notpresent'):
                return f"{root_nae} not present"
            elif column.endswith('_good'):
                return f"{root_nae} good"
            elif column.endswith('_poor'):
                return f"{root_nae} poor"
            else:
                return root_nae
    return column 

full_name_map = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class',
    'rbcc':'red blood cell count',
    'rbc': 'red blood cells',
    'ba': 'bacteria',
    'htn':'hypertension',
    
}

encode_dum_columns_QL_merge.rename(columns=\
{column: change_to_full_name(column) for column in encode_dum_columns_QL_merge.columns}
, inplace=True)
display(encode_dum_columns_QL_merge.columns)
display(encode_dum_columns_QL_merge)

Index(['age', 'blood pressure', 'blood glucose random', 'blood urea',
       'serum creatinine', 'sodium', 'potassium', 'hemoglobin',
       'packed cell volume', 'white blood cell count', 'red blood cell count',
       'albumin', 'sugar', 'red blood cells Abnormal', 'rbc_normal',
       'pus cell Abnormal', 'pc_normal', 'pus cell clumps not present',
       'pus cell clumps present', 'bacteria not present', 'bacteria present',
       'hypertension No', 'hypertension Yes', 'diabetes mellitus No',
       'diabetes mellitus Yes', 'coronary artery disease No',
       'coronary artery disease Yes', 'appetite good', 'appetite poor',
       'pedal edema No', 'pedal edema Yes', 'anemia No', 'anemia Yes',
       'Target_ckd', 'Target_notckd'],
      dtype='object')

,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,white blood cell count,...,coronary artery disease No,coronary artery disease Yes,appetite good,appetite poor,pedal edema No,pedal edema Yes,anemia No,anemia Yes,Target_ckd,Target_notckd
0,0.783784,0.00,0.154506,0.111111,0.037037,1.000000,0.114286,0.721311,0.684211,0.306452,...,1,0,1,0,1,0,1,0,0,1
1,0.472973,0.50,0.223176,0.098039,0.049383,0.600000,0.600000,0.942623,0.657895,0.387097,...,1,0,1,0,1,0,1,0,0,1
2,0.486486,0.25,0.098712,0.143791,0.061728,0.766667,0.514286,0.901639,0.710526,0.225806,...,1,0,1,0,1,0,1,0,0,1
3,0.918919,0.00,0.077253,0.261438,0.024691,0.900000,0.228571,0.950820,0.973684,0.137097,...,1,0,1,0,1,0,1,0,0,1
4,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,0.565574,0.552632,0.427419,...,0,1,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,0.605263,0.290323,...,1,0,1,0,1,0,1,0,1,0
130,0.324324,0.50,0.081545,0.209150,0.012346,0.633333,0.600000,0.909836,0.947368,0.475806,...,1,0,1,0,1,0,1,0,0,1
131,0.878378,0.25,0.047210,0.052288,0.049383,0.833333,0.600000,0.745902,0.736842,0.443548,...,1,0,1,0,1,0,1,0,0,1
132,0.851351,0.25,0.618026,0.562092,0.728395,0.000000,0.285714,0.311475,0.315789,0.580645,...,0,1,1,0,0,1,0,1,1,0


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM best_hes_man
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [13]:
import pandas as pd

best_hes_man = pd.DataFrame(encode_dum_columns_QL_merge)

# Group by 'Target' and count occurrences
result = best_hes_man.groupby('Target_ckd').size().reset_index(name='count')

# Display the result
print(result)

   Target_ckd  count
0           0    111
1           1     23


*Describe what the above code does here*

Basically, the 383GPT groups the DataFrame by column "Target_ckd". Then, it counts the frequencies in each group of "Target_ckd", i.e if the patient are ckd, then return 1; else, return 0. Finally, the algorithm reset index of the data set, and add one more column name 'count' to store occurencies. 